File: src/new/core/gui/CheckBox.ts
import { create_el } from "./dom";
import { WritableProperty } from "../observable/WritableProperty";
import { property } from "../observable";
import { LabelledControl } from "./LabelledControl";

export class CheckBox extends LabelledControl {
    readonly element: HTMLInputElement = create_el("input", "core_CheckBox");

    readonly checked: WritableProperty<boolean> = property(false);

    readonly preferred_label_position = "right";

    constructor(checked: boolean = false, label?: string) {
        super(label);

        this.element.type = "checkbox";
        this.element.onchange = () => this.checked.set(this.element.checked);

        this.disposables(
            this.checked.observe(checked => (this.element.checked = checked)),

            this.enabled.observe(enabled => (this.element.disabled = !enabled)),
        );

        this.checked.set(checked);
    }
}

File: webui/src/main/kotlin/world/phantasmal/webui/widgets/Checkbox.kt
package world.phantasmal.webui.widgets

import org.w3c.dom.Node
import world.phantasmal.observable.cell.Cell
import world.phantasmal.observable.cell.nullCell
import world.phantasmal.observable.cell.trueCell
import world.phantasmal.webui.dom.input

class Checkbox(
    visible: Cell<Boolean> = trueCell(),
    enabled: Cell<Boolean> = trueCell(),
    tooltip: Cell<String?> = nullCell(),
    label: String? = null,
    labelCell: Cell<String>? = null,
    preferredLabelPosition: LabelPosition = LabelPosition.After,
    private val checked: Cell<Boolean>? = null,
    private val onChange: ((Boolean) -> Unit)? = null,
) : LabelledControl(visible, enabled, tooltip, label, labelCell, preferredLabelPosition) {
    override fun Node.createElement() =
        input {
            id = labelId
            className = "pw-checkbox"
            type = "checkbox"

            if (this@Checkbox.checked != null) {
                observe(this@Checkbox.checked) {
                    checked = it
                }
            }

            if (onChange != null) {
                onchange = { onChange.invoke(checked) }
            }
        }
}
